smartsim/_core/mli/workermanager.py

```diff
@@ -0,0 +1,712 @@
+
+            # allow service to shutdown if no cooldown period applies...
+            running = not eligible_to_quit
+
+            # ... but verify we don't have remaining cooldown time
+            if self._cooldown:
+                elapsed_cooldown += start_ts - last_ts
+                remaining = cooldown_ns - elapsed_cooldown
+                running = remaining > 0
+
+                rem_in_s = remaining / nanosecond_scale_factor
+
+                if not running:
+                    cd_in_s = cooldown_ns / nanosecond_scale_factor
+                    logger.info(f"cooldown {cd_in_s}s exceeded by {abs(rem_in_s):.2f}s")
+                    self._on_cooldown()
+                    continue
+
+                logger.debug(f"cooldown remaining {abs(rem_in_s):.2f}s")
+
+            last_ts = start_ts
+            start_ts = time.time_ns()
+            time.sleep(1)
+
+        self._on_shutdown()
+
+
+class WorkerManager(ServiceHost):
+    """An implementation of a service managing distribution of tasks to
+    machine learning workers"""
+
+    def __init__(
+        self,
+        feature_store: FeatureStore,
+        worker: MachineLearningWorkerBase,
+        as_service: bool = False,
+        cooldown: int = 0,
+    ) -> None:
+        """Initialize the WorkerManager
+        :param feature_store: The persistence mechanism
+        :param worker: A worker to manage
+        :param as_service: Specifies run-once or run-until-complete behavior of service
+        :param cooldown: Number of seconds to wait before shutting down afer
+        shutdown criteria are met"""
+        super().__init__(as_service, cooldown)
+
+        self._workers: t.Dict[
+            str, "t.Tuple[MachineLearningWorkerBase, mp.Queue[bytes]]"
+        ] = {}
+        """a collection of workers the manager is controlling"""
+        self._upstream_queue: t.Optional[mp.Queue[bytes]] = None
+        """the queue the manager monitors for new tasks"""
+        self._feature_store: FeatureStore = feature_store
+        """a feature store to retrieve models from"""
+        self._worker = worker
+        """The ML Worker implementation"""
+
+    @property
+    def upstream_queue(self) -> "t.Optional[mp.Queue[bytes]]":
+        """Return the queue used by the worker manager to receive new work"""
+        return self._upstream_queue
+
+    @upstream_queue.setter
+    def upstream_queue(self, value: "mp.Queue[bytes]") -> None:
+        """Set/update the queue used by the worker manager to receive new work"""
+        self._upstream_queue = value
+
+    def _on_iteration(self, timestamp: int) -> None:
+        """Executes calls to the machine learning worker implementation to complete
+        the inference pipeline"""
+        logger.debug(f"{timestamp} executing worker manager pipeline")
+
+        if self.upstream_queue is None:
+            logger.warning("No queue to check for tasks")
+            return
+
+        msg: bytes = self.upstream_queue.get()
```

Comment:
After looking through the ``on_iteration`` loop, I think the queue should be the first part where we add abstractions and try to build in basic structures that will enable concurrent and parallel processing (e.g. grabbing more than one request through the loop at a time).  I'm going to pause the review and we can discuss as a team tomorrow.

Reply:
Created ticket 711 to follow-up on this task and batch request properties